sc_cost_meter/utils.py

```diff
@@ -0,0 +1,208 @@
+          therefore the response from get_cost_and_usage API may lag up to one day.
+          This means that the only way to get consistent billing info is to query the
+          previous day's billing.
+  '''
+  client = get_ce_client()
+
+  # Get the cost from the previous day
+  current_time = datetime.utcnow()
+  start_time = current_time - timedelta(days=2)
+  end_time = current_time - timedelta(days=1)
+
+  result = client.get_cost_and_usage(
+    TimePeriod={
+      "Start": start_time.strftime('%Y-%m-%d'),
+      "End": end_time.strftime('%Y-%m-%d')
+    },
+    Granularity="DAILY",
+    Filter={
+      "And": [
+        {
+          "Dimensions": {
+            "Key": "SERVICE",
+            "Values": [
+              "Amazon Elastic Compute Cloud - Compute",  # instance cost
+              "EC2 - Other"   # EB Volume, data transfer, elastic IP, etc..
+            ]
+          }
+        },
+        {
+          "Tags": tags
+        },
+      ]
+    },
+    Metrics=["UnblendedCost"]
+  )
+
+  cost = result[0]["Total"]["UnblendedCost"]["Amount"]   # total cost (EC2 instance + EC2-other)
+
+  return cost
+
+
+def get_tags(instance):
+  '''
+  Get all tags for an instance
+  :param instance: EC2 instance
+  :return: list of tags
+  '''
+  client = get_ec2_client()
+  response = client.describe_tags(
+    Filters = [
+       {
+          'Name': 'resource-id',
+          'Values': [
+             instance,
+          ]
+       }
+    ]
+  )
+  tags = response['Tags']
+  log.debug(f'instance tags = {tags}')
+  return tags
+
+
+def get_instance_product_code(instance):
+  '''
+  Get the Marketplace product code from the instance
+  :param instance: an EC2 instance
+  :return: product code
+  '''
+  product_codes = instance['ProductCodes']
+  if len(product_codes) >= 1:
+    for product_code in product_codes:
+       if product_code['ProductCodeType'] == 'marketplace':
+         product_code_id = product_code['ProductCodeId']
+         log.debug(f'Product Code ID = {product_code_id}')
+         return product_code_id
+
+  return None
+
+def get_marketplace_product_code(tags):
+  '''
+  Get the Marketplace product code for the Service Catalog SaaS service
+  that we apply as a tag to the resource when we provision it.
+  :param tags: list of tags
+  :return: Marketplace product code if a tag exists, otherwise None
+  '''
+  product_code = None
+  for tag in tags:
+    if tag['Key'] == 'marketplace:productCode':
+      product_code = tag['Value']
+
+  return product_code
+
+def report_usage(cost, customer_id, product_code):
+  '''
+  Report cost information to the Marketplace
+  :param cost: the product cost
+  :param customer_id: the customer identifier
+  :param product_code: the product code
+  :return: reporting status ('Success'|'CustomerNotSubscribed'|'DuplicateRecord')
+  '''
+  cost_accrued_rate = 0.001  # TODO: use mareketplace get_dimension API to get this info
```

Comment:
I think we need to do this before merging the PR:  Originally the rate was .001 and I asked for it to be changed to 1.0, then I asked for it to be changed back.  Other change requests have been slow to apply.  I am afraid that AWS will see the old / out-of-date request to change .001->1.0 and apply it.  Then we will accidentally bill (ourselves) for 1000 times too much.  Although we would be collecting money from ourselves, I've heard that AWS adds a substantial margin to the amount billed (in addition to any taxes) so we could accidentally blow hundreds of dollars or more.  The safe guard is to look up the actual rate rather than assuming a certain value.